### PR TITLE
ViewerWidget accepts TrajectoryData, and other improvements

### LINKED ISSUE #158: Jupyter Notebook: Widget Python Improvements
## Use Case
As a user, I want to be able to easily initialize, update, and save a simularium trajectory in a Jupyter notebook.

## Acceptance Criteria
1.  It is a relatively straightforward to provide the trajectory data to the viewer widget, with minimal simulariumio manipulations. This may look like being able to provide a regular `TrajectoryData` object to the widget, or providing all of the data pieces as parameters to the widget.
2. There are functions to update individual parts of the trajectory without writing out the entire trajectory data object again (ie just call `update_agent_data()` and provide new radii will just update the radii and retain all TrajectoryData parameters)
3. There is a function to save the trajectory represented in the widget locally, so the user can bring that into the simularium website if they desire.

### PR DESCRIPTION
**Pull request by ascibisz**
### Overview

Rather than expecting the user of the `ViewerWidget` to use the simulariumio `TrajectoryConverter` to convert their `TrajectoryData` object to a JSON string to send as an input to the widget, let's do that behind the scenes in the widget itself.

In addition, rather than type out the entire `TrajectoryData` object a second time to update something small, we provide various `update` functions that the user can call on the Widget instance to change various pieces of the trajectory. 

Finally, add in `.save("whatever_name")` to allow users to save their final trajectory locally, which is handy if they want to bring this to the full simularium viewer.

#### Question 1
I don't think I've found the sweet spot for naming the functions and the individual parameters for the various update functions. It's just weird because there are SO many parameters that can be updated, and since some are nested in different object types I'm not sure what the best to intuitively indicate what parameter we're referencing.

For context, here's how `TrajectoryData` objects are organized:
```
TrajectoryData(
    meta_data=MetaData(
        box_size: np.array,
        camera_defaults: CameraData(
            position: np.ndarray,
            look_at_position: np.ndarray,
            up_vector: np.ndarray,
            fov_degrees: float,
        ),
        scale_factor: float,
        trajectory_title: str,
        model_meta_data: ModelMetaData(
            title: str,
            version: str,
            authors: str,
            description: str,
            doi: str,
            source_code_url: str,
            source_code_license_url: str,
            input_data_url: str,
            raw_output_data_url: str,
        ),
    ),
    agent_data=AgentData(
        times: np.ndarray,
        n_agents: np.ndarray,
        viz_types: np.ndarray,
        unique_ids: np.ndarray,
        types: List,
        positions: np.ndarray,
        radii: np.ndarray,
        rotations: np.ndarray,
        n_subpoints: np.ndarray,
        subpoints: np.ndarray,
        display_data: Dict[str, DisplayData],
        draw_fiber_points: bool,
    ),
    time_units=UnitData(name: str, magnitude: float),
    spatial_units=UnitData(name: str, magnitude: float),
)
```
And here are the update functions / parameter names:
* **update_agent_data**(
        times: np.ndarray,
        n_agents: np.ndarray,
        viz_types: np.ndarray,
        unique_ids: np.ndarray,
        types: List,
        positions: np.ndarray,
        radii: np.ndarray,
        rotations: np.ndarray,
        n_subpoints: np.ndarray,
        subpoints: np.ndarray,
        display_data: Dict[str, DisplayData],
     )
* **update_meta_data**(
        box_size: np.ndarray,
        camera_position: np.ndarray,
        camera_look_at_position: np.ndarray,
        camera_up_vector: np.ndarray,
        camera_fov_degrees: float,
        scale_factor: float,
        trajectory_title: str,
        title: str,
        version: str,
        authors: str,
        description: str,
        source_code_url: str,
        input_data_url: str,
        raw_output_data_url: str,
    )
* **update_time_units**(name: str, magnitude: float)
* **update_spatial_units**(name: str, magnitude: float)

I also included a function to add additional display data without wiping out the entries that existed before. I don't know if this is actually useful or is making things more confusing. I just felt like there's a use case when you have already set display data for one agent and you want to try adding it for another, without deleting the first. We could also just make this the behavior that will happen in `update_agent_data` if the `display_data` arg is included.
* **add_display_data**(display_data: Dict[str, DisplayData])

#### Question 2
Do we want to have any other potential ways to send in the initial trajectory data? Certainly sending a `TrajectoryData` object in is better than having to convert it to a string, but I know Megan had talked about sending all of the parameters in individually, something like:
```
ViewerWidget(
   box_size=[],
   camera_defaults_position=[],
   look_at_position=[],
   fov_degrees=75.0,
   trajectory_title="",
   title="",
   version="",
   authors="",
   description="",
   ....
   times=[],
   n_agents=[],
   viz_types=[],
   unique_ids=[],
   types=[],
   ...
)
```

### Screenshots

Screen Shot 2023-09-15 at 10 46 39 AM

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -30,6 +30,7 @@ dependencies = [
     "jupyter>=1.0.0",
     "jupyterlab>=3.5.2",
     "numpy>=1.14.3",
+    "simulariumio",
```
Comment (meganrm) [inline pyproject.toml:33]: you could remove it from "examples" dependancies now

File: nbsv/viewer.py
Comment (meganrm) [inline nbsv/viewer.py:1]: this looks great

**Comment by meganrm** [comment]
Question1: Are you still thinking of folding time and space units into the meta data in simulariumio?
Question2: I think doing it this way is good

**Review by interim17 — APPROVED** [review]

**Review by toloudis — APPROVED** [review]

**Comment by ascibisz** [comment]
> Question1: Are you still thinking of folding time and space units into the meta data in simulariumio?

@meganrm I'm going to address that potentially in a simulariumio ticket that looks at reorganizing our data structures. I'll be sure to add something there to make sure this work stays compatible

**Review by meganrm — APPROVED** [review]

**Pull request merged**